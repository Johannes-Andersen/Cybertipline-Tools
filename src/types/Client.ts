/**
 * Generic wrapper for API responses that the CyberTipline Client returns
 */
export interface ResponseWrapper<T> {
  /** The wrapped response data */
  data: T;
  /**
   * Request ID from NCMEC's API. Used for troubleshooting,
   *    *
   * @see https://report.cybertip.org/ispws/documentation/index.html#request-ids
   */
  requestId: string;
}

/**
 * Environment options for the CyberTipline API
 */
export enum Environment {
  /** Production environment - https://report.cybertip.org/ispws/documentation/index.html */
  Production = 'production',
  /** Testing environment - https://exttest.cybertip.org/ispws/documentation/index.html */
  Testing = 'testing',
}

/**
 * Configuration options for the CyberTipline client
 */
export interface ClientConfig {
  /** The environment to use (production or testing) */
  environment: Environment;
  /** API credentials provided by NCMEC */
  credentials: {
    /** * Username provided by NCMEC */
    username: string;
    /** Password provided by NCMEC */
    password: string;
  };
}
