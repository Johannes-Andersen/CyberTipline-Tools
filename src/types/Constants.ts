/**
 * The response code returned from the submittal. 0 is success, any non-zero number is an error code.
 *
 * The following is a non-exhaustive list of possible error codes.
 * @see https://report.cybertip.org/ispws/documentation/index.html#report-response
 */
export enum ResponseCode {
  /** Success */
  Success = 0,
  /** Server error */
  ServerError = 1000,
  /** Save failed */
  SaveFailed = 1100,
  /** Upload failed */
  UploadFailed = 1110,
  /** File upload failed */
  FileUploadFailed = 1111,
  /** Resource not found */
  ResourceNotFound = 1210,
  /** Update failed */
  UpdateFailed = 1300,
  /** Authentication required */
  AuthenticationRequired = 2000,
  /** User is not authorized to perform this action */
  NotAuthorized = 3000,
  /** User does not have authorization to make submissions */
  NotAuthorizedForSubmissions = 3100,
  /** User does not have authorization to make updates */
  NotAuthorizedForUpdates = 3300,
  /** Invalid request */
  InvalidRequest = 4000,
  /** Validation failed */
  ValidationFailed = 4100,
  /** Malformed XML submittal */
  MalformedXmlSubmittal = 4110,
  /** Malformed file submittal */
  MalformedFileSubmittal = 4200,
  /** Report does not exist */
  ReportDoesNotExist = 5001,
  /** File does not exist */
  FileDoesNotExist = 5002,
  /** Report already retracted */
  ReportAlreadyRetracted = 5101,
  /** Report already finished */
  ReportAlreadyFinished = 5102,
}

export enum IncidentType {
  ChildPornography = 'Child Pornography (possession, manufacture, and distribution)',
  ChildSexTrafficking = 'Child Sex Trafficking',
  ChildSexTourism = 'Child Sex Tourism',
  ChildSexualMolestation = 'Child Sexual Molestation',
  MisleadingDomainName = 'Misleading Domain Name',
  MisleadingWordsOrImages = 'Misleading Words or Digital Images on the Internet',
  OnlineEnticement = 'Online Enticement of Children for Sexual Act',
  UnsolicitedObsceneMaterial = 'Unsolicited Obscene Material Sent to a Child',
}

/**
 * Tags to describe the report.
 * @see https://report.cybertip.org/ispws/documentation/index.html#report-annotations
 */
export enum ReportAnnotation {
  /** The report is associated with sextortion. */
  Sextortion = 'sextortion',
  /** The report is associated with solicitation of child sexual assault material (CSAM). */
  CsamSolicitation = 'csamSolicitation',
  /** The report is associated with an interaction between minors. */
  MinorToMinorInteraction = 'minorToMinorInteraction',
  /** The report is associated with spam. */
  Spam = 'spam',
}

/**
 * Tags used to describe a file.
 *
 * @see https://report.cybertip.org/ispws/documentation/index.html#file-annotations
 */
export enum FileAnnotations {
  /** The file is depicting anime, drawing, cartoon, virtual or hentai. */
  AnimeDrawingVirtualHentai = 'animeDrawingVirtualHentai',
  /** The file is being shared/posted out of mimicry or other seemingly non-malicious intent. */
  PotentialMeme = 'potentialMeme',
  /** The file is circulating rapidly from one user to another. */
  Viral = 'viral',
  /** The file contains content that is believed to be self-produced */
  PossibleSelfProduction = 'possibleSelfProduction',
  /** The file depicts an intentional act of causing physical injury or trauma to a person. */
  PhysicalHarm = 'physicalHarm',
  /** The file depicts graphic violence, including but not limited to acts of brutality or detailed or vivid gruesomeness. */
  ViolenceGore = 'violenceGore',
  /** The file involves an animal. */
  Bestiality = 'bestiality',
  /** The file depicts content that was streamed live at the time it was uploaded. */
  LiveStreaming = 'liveStreaming',
  /** The file depicts an infant. */
  Infant = 'infant',
  /** The file contains content that is believed to be Generative Artificial Intelligence. */
  GenerativeAI = 'generativeAi',
}

export enum FileRelevance {
  /** Content that is the motivation for making the CyberTipline report, according to the chosen incident type.
   *
   * Only "Reported" files may be identified as potential meme or be given an industry classification.
   */
  Reported = 'Reported',
  /** Supplementary content that provides contextual value to the report.
   *
   * A part of the communication containing apparent child pornography, including any data or information regarding the transmission of the communication
   * or any images, data, or other digital files contained in, or attached to, the communication (18 U.S. Code ยง 2258A (b)(5)).
   */
  SupplementalReported = 'Supplemental Reported',
}

/**
 * A = Prepubescent Minor
 * B = Pubescent Minor
 *
 * 1 (Sex Act): Any imagery depicting sexual intercourse (
 * including genital-genital, oral-genital, anal-genital,
 * or oral-anal whether between person of the same or opposite sex),
 * bestiality, masturbation, sadistic or masochistic abuse, degradation,
 * or any such depiction of the above that lacks serious literary, artistic, political, or scientific value.
 *
 * 2 (Lascivious Exhibition): Any imagery depicting the lascivious exhibition of the anus, genitals, or pubic area of any person,
 * where a minor is engaging in the lascivious exhibition or being used in connection with sexually explicit conduct,
 * which may include but is not limited to imagery where the focal point is on the child's anus, genitals,
 * or pubic area and where the depiction is intended or designed to elicit a sexual response in the viewer.
 */
export enum IndustryClassification {
  /** Prepubescent Minor - Sex Act*/
  A1 = 'A1',
  /** Prepubescent Minor - Lascivious Exhibition */
  A2 = 'A2',
  /** Pubescent Minor - Sex Act */
  B1 = 'B1',
  /** Pubescent Minor - Lascivious Exhibition */
  B2 = 'B2',
}
