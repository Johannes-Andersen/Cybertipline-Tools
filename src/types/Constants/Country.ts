/**
 * ISO 3166-1 alpha-2 two-character country code.
 */
export enum Country {
  // Specials
  AnonymousProxy = 'A1',
  SatelliteProvider = 'A2',
  OtherCountry = 'O1',
  // Country codes
  Andorra = 'AD',
  UnitedArabEmirates = 'AE',
  Afghanistan = 'AF',
  AntiguaAndBarbuda = 'AG',
  Anguilla = 'AI',
  Albania = 'AL',
  Armenia = 'AM',
  NetherlandsAntilles = 'AN',
  Angola = 'AO',
  AsiaPacificRegion = 'AP',
  Antarctica = 'AQ',
  Argentina = 'AR',
  AmericanSamoa = 'AS',
  Austria = 'AT',
  Australia = 'AU',
  Aruba = 'AW',
  AlandIslands = 'AX',
  Azerbaijan = 'AZ',
  BosniaAndHerzegovina = 'BA',
  Barbados = 'BB',
  Bangladesh = 'BD',
  Belgium = 'BE',
  BurkinaFaso = 'BF',
  Bulgaria = 'BG',
  Bahrain = 'BH',
  Burundi = 'BI',
  Benin = 'BJ',
  SaintBartelemey = 'BL',
  Bermuda = 'BM',
  BruneiDarussalam = 'BN',
  Bolivia = 'BO',
  BonaireSintEustatiusAndSaba = 'BQ',
  Brazil = 'BR',
  Bahamas = 'BS',
  Bhutan = 'BT',
  BouvetIsland = 'BV',
  Botswana = 'BW',
  Belarus = 'BY',
  Belize = 'BZ',
  Canada = 'CA',
  CocosKeelingIslands = 'CC',
  CongoDemocraticRepublic = 'CD',
  CentralAfricanRepublic = 'CF',
  Congo = 'CG',
  Switzerland = 'CH',
  CoteDIvoire = 'CI',
  CookIslands = 'CK',
  Chile = 'CL',
  Cameroon = 'CM',
  China = 'CN',
  Colombia = 'CO',
  CostaRica = 'CR',
  Cuba = 'CU',
  CapeVerde = 'CV',
  ChristmasIsland = 'CX',
  Curacao = 'CW',
  Cyprus = 'CY',
  CzechRepublic = 'CZ',
  Germany = 'DE',
  Djibouti = 'DJ',
  Denmark = 'DK',
  Dominica = 'DM',
  DominicanRepublic = 'DO',
  Algeria = 'DZ',
  Ecuador = 'EC',
  Estonia = 'EE',
  Egypt = 'EG',
  WesternSahara = 'EH',
  Eritrea = 'ER',
  Spain = 'ES',
  Ethiopia = 'ET',
  Europe = 'EU',
  Finland = 'FI',
  Fiji = 'FJ',
  FalklandIslands = 'FK',
  Micronesia = 'FM',
  FaroeIslands = 'FO',
  France = 'FR',
  FranceMetropolitan = 'FX',
  Gabon = 'GA',
  UnitedKingdom = 'GB',
  Grenada = 'GD',
  Georgia = 'GE',
  FrenchGuiana = 'GF',
  Guernsey = 'GG',
  Ghana = 'GH',
  Gibraltar = 'GI',
  Greenland = 'GL',
  Gambia = 'GM',
  Guinea = 'GN',
  Guadeloupe = 'GP',
  EquatorialGuinea = 'GQ',
  Greece = 'GR',
  SouthGeorgia = 'GS',
  Guatemala = 'GT',
  Guam = 'GU',
  GuineaBissau = 'GW',
  Guyana = 'GY',
  HongKong = 'HK',
  HeardIsland = 'HM',
  Honduras = 'HN',
  Croatia = 'HR',
  Haiti = 'HT',
  Hungary = 'HU',
  Indonesia = 'ID',
  Ireland = 'IE',
  Israel = 'IL',
  IsleOfMan = 'IM',
  India = 'IN',
  BritishIndianOceanTerritory = 'IO',
  Iraq = 'IQ',
  Iran = 'IR',
  Iceland = 'IS',
  Italy = 'IT',
  Jersey = 'JE',
  Jamaica = 'JM',
  Jordan = 'JO',
  Japan = 'JP',
  Kenya = 'KE',
  Kyrgyzstan = 'KG',
  Cambodia = 'KH',
  Kiribati = 'KI',
  Comoros = 'KM',
  SaintKittsAndNevis = 'KN',
  NorthKorea = 'KP',
  SouthKorea = 'KR',
  Kuwait = 'KW',
  CaymanIslands = 'KY',
  Kazakhstan = 'KZ',
  Laos = 'LA',
  Lebanon = 'LB',
  SaintLucia = 'LC',
  Liechtenstein = 'LI',
  SriLanka = 'LK',
  Liberia = 'LR',
  Lesotho = 'LS',
  Lithuania = 'LT',
  Luxembourg = 'LU',
  Latvia = 'LV',
  Libya = 'LY',
  Morocco = 'MA',
  Monaco = 'MC',
  Moldova = 'MD',
  Montenegro = 'ME',
  SaintMartin = 'MF',
  Madagascar = 'MG',
  MarshallIslands = 'MH',
  Macedonia = 'MK',
  Mali = 'ML',
  Myanmar = 'MM',
  Mongolia = 'MN',
  Macao = 'MO',
  NorthernMarianaIslands = 'MP',
  Martinique = 'MQ',
  Mauritania = 'MR',
  Montserrat = 'MS',
  Malta = 'MT',
  Mauritius = 'MU',
  Maldives = 'MV',
  Malawi = 'MW',
  Mexico = 'MX',
  Malaysia = 'MY',
  Mozambique = 'MZ',
  Namibia = 'NA',
  NewCaledonia = 'NC',
  Niger = 'NE',
  NorfolkIsland = 'NF',
  Nigeria = 'NG',
  Nicaragua = 'NI',
  Netherlands = 'NL',
  Norway = 'NO',
  Nepal = 'NP',
  Nauru = 'NR',
  Niue = 'NU',
  NewZealand = 'NZ',
  Oman = 'OM',
  Panama = 'PA',
  Peru = 'PE',
  FrenchPolynesia = 'PF',
  PapuaNewGuinea = 'PG',
  Philippines = 'PH',
  Pakistan = 'PK',
  Poland = 'PL',
  SaintPierreAndMiquelon = 'PM',
  Pitcairn = 'PN',
  PuertoRico = 'PR',
  PalestinianTerritory = 'PS',
  Portugal = 'PT',
  Palau = 'PW',
  Paraguay = 'PY',
  Qatar = 'QA',
  Reunion = 'RE',
  Romania = 'RO',
  Serbia = 'RS',
  Russia = 'RU',
  Rwanda = 'RW',
  SaudiArabia = 'SA',
  SolomonIslands = 'SB',
  Seychelles = 'SC',
  Sudan = 'SD',
  Sweden = 'SE',
  Singapore = 'SG',
  SaintHelena = 'SH',
  Slovenia = 'SI',
  SvalbardAndJanMayen = 'SJ',
  Slovakia = 'SK',
  SierraLeone = 'SL',
  SanMarino = 'SM',
  Senegal = 'SN',
  Somalia = 'SO',
  Suriname = 'SR',
  SouthSudan = 'SS',
  SaoTomeAndPrincipe = 'ST',
  ElSalvador = 'SV',
  SintMaarten = 'SX',
  Syria = 'SY',
  Swaziland = 'SZ',
  TurksAndCaicosIslands = 'TC',
  Chad = 'TD',
  FrenchSouthernTerritories = 'TF',
  Togo = 'TG',
  Thailand = 'TH',
  Tajikistan = 'TJ',
  Tokelau = 'TK',
  TimorLeste = 'TL',
  Turkmenistan = 'TM',
  Tunisia = 'TN',
  Tonga = 'TO',
  Turkey = 'TR',
  TrinidadAndTobago = 'TT',
  Tuvalu = 'TV',
  Taiwan = 'TW',
  Tanzania = 'TZ',
  Ukraine = 'UA',
  Uganda = 'UG',
  UnitedStatesMinorOutlyingIslands = 'UM',
  UnitedStates = 'US',
  Uruguay = 'UY',
  Uzbekistan = 'UZ',
  VaticanCity = 'VA',
  SaintVincentAndTheGrenadines = 'VC',
  Venezuela = 'VE',
  BritishVirginIslands = 'VG',
  UsVirginIslands = 'VI',
  Vietnam = 'VN',
  Vanuatu = 'VU',
  WallisAndFutuna = 'WF',
  Samoa = 'WS',
  Kosovo = 'XK',
  Yemen = 'YE',
  Mayotte = 'YT',
  SouthAfrica = 'ZA',
  Zambia = 'ZM',
  Zimbabwe = 'ZW',
}
