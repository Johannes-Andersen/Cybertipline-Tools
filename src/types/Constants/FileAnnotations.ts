/**
 * Tags used to describe a file.
 *
 * @see https://report.cybertip.org/ispws/documentation/index.html#file-annotations
 */
export enum FileAnnotations {
  /** The file is depicting anime, drawing, cartoon, virtual or hentai. */
  AnimeDrawingVirtualHentai = 'animeDrawingVirtualHentai',
  /** The file is being shared/posted out of mimicry or other seemingly non-malicious intent. */
  PotentialMeme = 'potentialMeme',
  /** The file is circulating rapidly from one user to another. */
  Viral = 'viral',
  /** The file contains content that is believed to be self-produced */
  PossibleSelfProduction = 'possibleSelfProduction',
  /** The file depicts an intentional act of causing physical injury or trauma to a person. */
  PhysicalHarm = 'physicalHarm',
  /** The file depicts graphic violence, including but not limited to acts of brutality or detailed or vivid gruesomeness. */
  ViolenceGore = 'violenceGore',
  /** The file involves an animal. */
  Bestiality = 'bestiality',
  /** The file depicts content that was streamed live at the time it was uploaded. */
  LiveStreaming = 'liveStreaming',
  /** The file depicts an infant. */
  Infant = 'infant',
  /** The file contains content that is believed to be Generative Artificial Intelligence. */
  GenerativeAI = 'generativeAi',
}
