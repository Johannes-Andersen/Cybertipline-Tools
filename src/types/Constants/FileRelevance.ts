export enum FileRelevance {
  /** Content that is the motivation for making the CyberTipline report, according to the chosen incident type.
   *
   * Only "Reported" files may be identified as potential meme or be given an industry classification.
   */
  Reported = 'Reported',
  /** Supplementary content that provides contextual value to the report.
   *
   * A part of the communication containing apparent child pornography, including any data or information regarding the transmission of the communication
   * or any images, data, or other digital files contained in, or attached to, the communication (18 U.S. Code ยง 2258A (b)(5)).
   */
  SupplementalReported = 'Supplemental Reported',
}
