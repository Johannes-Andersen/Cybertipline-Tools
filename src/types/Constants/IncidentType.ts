export enum IncidentType {
  ChildPornography = 'Child Pornography (possession, manufacture, and distribution)',
  ChildSexTrafficking = 'Child Sex Trafficking',
  ChildSexTourism = 'Child Sex Tourism',
  ChildSexualMolestation = 'Child Sexual Molestation',
  MisleadingDomainName = 'Misleading Domain Name',
  MisleadingWordsOrImages = 'Misleading Words or Digital Images on the Internet',
  OnlineEnticement = 'Online Enticement of Children for Sexual Act',
  UnsolicitedObsceneMaterial = 'Unsolicited Obscene Material Sent to a Child',
}
