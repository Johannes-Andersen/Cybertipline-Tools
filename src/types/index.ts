// Common types
export * from './common/Address.js';
export * from './common/DeviceId.js';
export * from './common/Email.js';
export * from './common/EstimatedLocation.js';
export * from './common/IpCapture.js';
export * from './common/Person.js';
export * from './common/Phone.js';

// File Details types
export * from './fileDetails/FileDetails.js';
export * from './fileDetails/OriginalFileHash.js';

// Report Incident types
export * from './incidents/CellPhoneIncident.js';
export * from './incidents/ChatImIncident.js';
export * from './incidents/EmailIncident.js';
export * from './incidents/NewsgroupIncident.js';
export * from './incidents/NonInternetIncident.js';
export * from './incidents/OnlineGamingIncident.js';
export * from './incidents/Peer2peerIncident.js';
export * from './incidents/WebPageIncident.js';

// Report Incident Details types
export * from './reportDetails/IncidentSummary.js';
export * from './reportDetails/IntendedRecipient.js';
export * from './reportDetails/LawEnforcement.js';
export * from './reportDetails/PersonOrUserReported.js';
export * from './reportDetails/Reporter.js';
export * from './reportDetails/Victim.js';

// Transactions types
export * from './transactions/CancelReport.js';
export * from './transactions/FinishReport.js';
export * from './transactions/GetStatus.js';
export * from './transactions/Report.js';
export * from './transactions/SubmitFileDetails.js';
export * from './transactions/UploadFile.js';

// Constants
export * from './Constants/Country.js';
export * from './Constants/FileAnnotations.js';
export * from './Constants/FileRelevance.js';
export * from './Constants/IncidentType.js';
export * from './Constants/IndustryClassification.js';
export * from './Constants/ReportAnnotation.js';
export * from './Constants/ResponseCode.js';
export * from './Constants/State.js';

// Client
export * from './Client.js';
