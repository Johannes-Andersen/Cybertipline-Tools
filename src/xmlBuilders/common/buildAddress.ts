import { XMLBuilder } from 'fast-xml-parser';
import type { Address } from '../../types';

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
});

export const buildAddress = (address: Address, keyName = 'address'): string =>
  builder.build({
    [keyName]: {
      '@_type': address.type,
      address: address.address,
      city: address.city,
      zipCode: address.zipCode,
      state: address.state,
      nonUsaState: address.nonUsaState,
      country: address.country,
    },
  });
