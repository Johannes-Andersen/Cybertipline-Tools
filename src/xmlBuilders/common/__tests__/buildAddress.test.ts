import { describe, expect, it } from 'vitest';
import { type Address, AddressType } from '../../../types';
import { Country } from '../../../types/Constants/Country';
import { State } from '../../../types/Constants/State';
import { buildAddress } from '../buildAddress';

describe('buildAddress', () => {
  it('should build an American address', () => {
    const address: Address = {
      address: '123 Main St',
      city: 'Anytown',
      state: State.Utah,
      country: Country.UnitedStates,
      zipCode: '12345',
      type: AddressType.Business,
    };

    const result = buildAddress(address);

    expect(result).toMatchSnapshot();
  });

  it('should build an European address', () => {
    const address: Address = {
      address: 'Calle de la Paz, 1',
      city: 'Madrid',
      nonUsaState: 'SomeRegion',
      country: Country.Spain,
      zipCode: '12345',
      type: AddressType.Business,
    };

    const result = buildAddress(address);

    expect(result).toMatchSnapshot();
  });
});
