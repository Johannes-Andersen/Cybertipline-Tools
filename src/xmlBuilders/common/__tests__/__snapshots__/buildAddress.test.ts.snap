// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildAddress > should build an American address 1`] = `"<address type="Business"><address>123 Main St</address><city>Anytown</city><zipCode>12345</zipCode><state>UT</state><country>US</country></address>"`;

exports[`buildAddress > should build an European address 1`] = `"<address type="Business"><address>Calle de la Paz, 1</address><city>Madrid</city><zipCode>12345</zipCode><nonUsaState>SomeRegion</nonUsaState><country>ES</country></address>"`;
